rust/src/rfb/rfb.rs

```diff
@@ -170,6 +179,10 @@ impl RFBState {
                             consumed += current.len() - rem.len();
                             current = rem;
 
+                            if let Some(frame) = &pdu {
```

Comment:
Should not there be a new frame at each iteration of the loop ?

Reply:
I am registering a PDU frame at the start of both `parse_request` and `parse_response` functions.
It is going to match either one of the match criteria and would go to the other function.

Reply:
But `parse_request`. does a loop that can create multiple transactions.

If so, there should me multiple frames, right ?

Reply:
Maybe I am still missing something but as far as what I have understood from the working of the function. Once it matches one of the states in `parse_request` the state changes. .e.g once it matches `TSClientProtocolVersion`, the state changes to `TCServerSecurityType` at line 188. once we have this new state. it will only match something in `parse_response` function and will not be parsed in the same in `parse_request`.

Once it goes to `parse_response` a new frame for pdu gets registered.

Reply:
So, does the parser need the `loop` at all ? ie Can it loop more than once ?

Reply:
I might need more insights into this as well

Reply:
Sorry for taking so long to answer this, I didn't realize it had a pending discussion going on. I am not familiar with RFB, but from what I can see, it seems that it expects to be able to parse more than one message at once - although I don't understand when would that happen. This seems necessary as of now due to the function logic, though, as we are not returning AppLayerResult::ok() when we successfully parse a message, but only once the input is totally consumed 🤔 

I think the simplest solution to this would be to register the frame as a whole once we have the Ok matches, maybe check how it's done with SMB - which also has it created inside a loop, only that there it's a `while`...

Reply:
So you are saying at every enum match for an RFB state, I should register the PDU frame instead of at the start?

Reply:
Also in this scenario I can directly specify the length instead of updating it later as the frame is getting at the same place now e.g
```
` parser::RFBGlobalState::TCServerInit => {
                    match parser::parse_server_init(current) {
                        Ok((rem, request)) => {
                            consumed += current.len() - rem.len();
                            current = rem;
```
**REGISTER FRAME HERE**
```
                            if let Some(frame) = &pdu {
                                frame.set_len(flow, consumed as i64); // THIS GETS REMOVED
                            }`
```

Reply:
Yes, I think we could try that approach, to prevent having the pdu overwritten in case the parser is indeed able to parse more than one message in one call. (I suspect the loop is there to ensure we have enough data to process one single message, but I'm not sure and the way it's implemented, IMO, leaves room for both scenarios, so better safe than sorry?)

Reply:
I think one call to the rfb or other app-layer parser should be able to register multiple PDU frames, right @victorjulien ?